bluemira/utilities/opt_problems.py

```diff
@@ -0,0 +1,249 @@
+    to a specific objective function for that subclass.
+
+    Parameters
+    ----------
+    parameterisation: any
+        Object storing parameterisation data to be optimised.
+    optimiser: Optimiser (default: None)
+        Optimiser object to use for constrained optimisation.
+        Does not need to be provided if not used by
+        optimise(), such as for purely unconstrained
+        optimisation.
+    objective: OptimisationObjective (default: None)
+        OptimisationObjective storing objective information to
+        provide to the Optimiser.
+    constraints: List[OptimisationConstraint]
+        Optional list of OptimisationConstraint objects storing
+        information about constraints that must be satisfied
+        during the optimisation, to be provided to the
+        Optimiser.
+    """
+
+    def __init__(
+        self,
+        parameterisation,
+        optimiser: Optimiser,
+        objective: OptimisationObjective,
+        constraints: List[OptimisationConstraint],
+    ):
+        self._parameterisation = parameterisation
+        self.opt = optimiser
+        self._objective = objective
+        self._constraints = constraints
+
+    def set_up_optimiser(self, dimension: int, bounds: np.array):
+        """
+        Set up NLOpt-based optimiser with algorithm,  bounds, tolerances, and
+        constraint & objective functions.
+
+        Parameters
+        ----------
+        dimension: int
+            Number of independent variables in the state vector to be optimised.
+        bounds: tuple
+            Tuple containing lower and upper bounds on the state vector.
+        """
+        # Build NLOpt optimiser, with optimisation strategy and length
+        # of state vector
+        self.opt.build_optimiser(n_variables=dimension)
+
+        # Apply objective function to optimiser
+        self._objective.apply_objective(self)
+
+        # Apply constraints
+        self.apply_constraints(self.opt, self._constraints)
+
+        # Set state vector bounds (current limits)
+        self.opt.set_lower_bounds(bounds[0])
+        self.opt.set_upper_bounds(bounds[1])
+
+    def apply_constraints(
+        self, opt: Optimiser, opt_constraints: List[OptimisationConstraint]
+    ):
+        """
+        Updates the optimiser in-place to apply problem constraints.
+        To be overidden by child classes to apply specific constraints.
+
+        Parameters
+        ----------
+        opt: Optimiser
+            Optimiser on which to apply the constraints. Updated in place.
+        opt_constraints: iterable
+            Iterable of OptimisationConstraint objects containing optimisation
+            constraints to be applied to the Optimiser.
+
+        Notes
+        -----
+        Lambda functions are used here to ensure the CoilsetPositionCOP is passed
+        to the function, to allow any properties that are stored in the
+        CoilsetPositionCOP to be accessed at runtime.
```

Comment:
Make less specific to equilibrium opt problems

Reply:
Updated